File: 30-Days_Data_Structures/Day-04_Recursion/Q01-Fibonacci_DP.js
"use strict";

const fibonacciDp = (num) => {
  const ht = {
    0: 0,
    1: 1,
  };
  if (num in ht) {
    return ht[num];
  } else {
    ht[num] = fibonacciDp(num - 1) + fibonacciDp(num - 2);
    return ht[num];
  }
};

fibonacciDp(4);

File: 30-Days_Data_Structures/Day-04_Recursion/Q01-Fibonacci_DP.py
def fibonacci_dp(num):
    ht = {
        0: 0,
        1: 1
    }

    if num in ht:
        return ht[num]
    else:
        ht[num] = fibonacci_dp(num - 1) + fibonacci_dp(num - 2)
        return ht[num]


print(fibonacci_dp(30))
